akka-projection-r2dbc/src/main/scala/akka/projection/r2dbc/internal/PostgresOffsetStoreDao.scala

```diff
@@ -338,6 +341,22 @@ private[projection] class PostgresOffsetStoreDao(
         .bindTimestamp(3, timestamp))
   }
 
+  override def adoptTimestampOffsets(latestBySlice: Seq[LatestBySlice]): Future[Long] = {
+    Future.sequence(latestBySlice.map(adoptTimestampOffset)).map(_.sum)
+  }
```

Comment:
This is just making the requests individually for now, for simplicity in creating queries. Could be batched together. Because this is targeting offsets directly by primary key, if combined that would be something like a Bitmap Or over Bitmap Index Scans.

Reply:
Thinking about if we could insert instead of update, using the batching mechanism that is in insertTimestampOffsetInTx? Then we would have to increase the timestamp with 1 microsecond to make the pk unique. Too hacky, I guess.

We don't expect this to be done that frequently so maybe not worry too much about performance? This code executes all in parallel, and that might be too hard on the connection pool? executor.updateInBatch, and using statement.add might be good enough. That is not very efficient batching because there is a roundtrip to db for each statement, but could be enough for this.

Reply:
> Then we would have to increase the timestamp with 1 microsecond to make the pk unique. Too hacky, I guess.

I think so. Useful to keep the original timestamps for debugging.

> We don't expect this to be done that frequently so maybe not worry too much about performance?

Yeah. At most, it would currently adopt an offset for every slice at some point, but then won't be triggered again.

Maybe batching works well if this becomes a scheduled maintenance task on the side. But need to do the triggers and state updates differently then. I'll take a look at this.

Reply:
Now that it's a separate scheduled task, more likely to have several updates at once. Have switched this to batched updates, with dynamically binding the OR conditions. This will be a bitmap heap scan and bitmap index scans for each query in Postgres, which should be more efficient than the separate queries and roundtrips.